scripts/globals/job_utils/beastmaster.lua

```diff
@@ -0,0 +1,132 @@
+-----------------------------------
+-- beastmaster Job Utilities
+-----------------------------------
+
+-----------------------------------
+xi = xi or {}
+xi.job_utils = xi.job_utils or {}
+xi.job_utils.beastmaster = xi.job_utils.beastmaster or {}
+
+local function getCharmDuration(charmer, target)
+    local charmDuration = 0
+
+    -- Calculate base duration and dLvl
+    local baseCharmDuration = math.floor(1.25 * charmer:getStat(xi.mod.CHR) + 150)
+    local dLvl = charmer:getMainLvl() - target:getMainLvl()
+
+    -- Default multiplier for dLvl -6 or lower
+    local dLvlCharmMult = 1 / 24
+
+    if dLvl >= -6 and dLvl < 9 then
+        -- Quintic fitting of duration multiplier as function of dLvl (r^2 > 0.999)
+        dLvlCharmMult = 0.9997336 + 0.3652882 * dLvl + 0.02097742 * dLvl^2
+                    - 0.004106429 * dLvl^3 + 0.000007231037 * dLvl^4
+                    + 0.00005102634 * dLvl^5
+    -- Caps at dLvl > 9
+    elseif dLvl >= 9 then
+        dLvlCharmMult = 6
+    end
+
+    -- Apply the dLvl multiplier
+    charmDuration = baseCharmDuration * dLvlCharmMult
+
+    -- Apply charm duration extension from gear
+    local charmTimeMod = charmer:getMod(xi.mod.CHARM_TIME)
+    local extraDurationFromMod = charmDuration * (charmTimeMod * 0.5 / 10) -- Assumes 5% per charmTimeMod
+    charmDuration = charmDuration + extraDurationFromMod
+
+    return math.floor(charmDuration)
+end
+
+xi.job_utils.beastmaster.getCharmChance = function(charmer, target, includeMods)
+    if
+        not charmer or                                -- Invalid charmer
+        not target or                                 -- Invalid target
+        not charmer:isPC() or                         -- Charmer not a player
+        not target:isMob() or                         -- Target not a mob
+        target:getMobMod(xi.mobMod.CHARMABLE) == 0 or -- Not charmable
+        (target:getMaster() ~= nil)                   -- Someone else's pet
```

Comment:
unneeded parenthesis

Reply:
Thanks for the feedback, I have fixed the PR.